packages/react-vapor/src/components/table-hoc/TableWithUrlState.tsx

```diff
@@ -0,0 +1,166 @@
+import * as QueryString from 'query-string';
+import * as React from 'react';
+import {connect} from 'react-redux';
+import * as _ from 'underscore';
+
+import {IReactVaporState} from '../../ReactVapor';
+import {callIfDefined} from '../../utils/FalsyValuesUtils';
+import {IDispatch, IThunkAction} from '../../utils/ReduxUtils';
+import {UrlUtils} from '../../utils/UrlUtils';
+import {applyDatePicker, changeDatePickerLowerLimit, changeDatePickerUpperLimit} from '../datePicker/DatePickerActions';
+import {filterThrough} from '../filterBox/FilterBoxActions';
+import {selectListBoxOption} from '../listBox/ListBoxActions';
+import {changePage} from '../navigation/pagination/NavigationPaginationActions';
+import {changePerPage} from '../navigation/perPage/NavigationPerPageActions';
+import {TableHeaderActions} from './actions/TableHeaderActions';
+import {ITableHOCOwnProps} from './TableHOC';
+import {ITableHOCPredicateValue, TableHOCUtils} from './TableHOCUtils';
+
+export interface TableWithUrlStateProps {
+    onUpdateUrl: (queryString: string) => void;
+}
+
+enum SortOrderValues {
+    ascending = 'asc',
+    descending = 'desc',
+}
+
+const Params = {
+    pageNumber: 'page',
+    pageSize: 'pageSize',
+    sortKey: 'sortBy',
+    sortOrder: 'order',
+    lowerDateLimit: 'from',
+    upperDateLimit: 'to',
+    filter: 'q',
+};
+
+export function tableWithUrlState<P extends ITableHOCOwnProps>(Component: React.ComponentType<P>) {
+    type Props = P &
+        TableWithUrlStateProps &
+        ReturnType<typeof mapStateToProps> &
+        ReturnType<typeof mapDispatchToProps>;
+
+    const mapStateToProps = (state: IReactVaporState, ownProps: P) => ({
+        query: getQuery(state, ownProps.id),
+    });
+
+    const mapDispatchToProps = (dispatch: IDispatch, ownProps: P) => ({
+        initializeFromUrl: () => dispatch(updateTableStateFromUrl(ownProps.id)),
+    });
+
+    class WrappedComponentDisconnected extends React.PureComponent<Props> {
+        static displayName = `withUrlState(${Component.displayName})`;
+
+        render() {
+            const wrappedProps = _.omit(this.props, 'onUpdate', 'onUpdateUrl', 'query', 'initializeFromUrl');
+            return <Component {...(wrappedProps as P)} onUpdate={this.onUpdate} />;
+        }
+
+        componentDidMount() {
+            this.props.initializeFromUrl();
+        }
+
+        private onUpdate = () => {
+            this.updateUrl(this.props.query);
+            callIfDefined(this.props.onUpdate);
+        };
+
+        private updateUrl = _.debounce(this.props.onUpdateUrl || _.noop, 100);
+    }
+
+    return connect(
+        mapStateToProps,
+        mapDispatchToProps
+        // @ts-ignore
```

Comment:
This was throwing an error?

Reply:
Yea, I spent some time trying to fix it without success. I find its better to ignore the error but still keep TypeScipt's type inference, rather then type as `any`.

```
class WrappedComponentDisconnected
Argument of type 'typeof WrappedComponentDisconnected' is not assignable to parameter of type 'ComponentType<Matching<{ query: string; } & { initializeFromUrl: any; }, Props>>'.
  Type 'typeof WrappedComponentDisconnected' is not assignable to type 'ComponentClass<Matching<{ query: string; } & { initializeFromUrl: any; }, Props>, any>'.
    Types of parameters 'props' and 'props' are incompatible.
      Type 'Matching<{ query: string; } & { initializeFromUrl: any; }, Props>' is not assignable to type 'Readonly<Props>'.
        Type 'P extends "query" | "initializeFromUrl" ? ({ query: string; } & { initializeFromUrl: any; })[P] extends Props[P] ? Props[P] : ({ query: string; } & { initializeFromUrl: any; })[P] : Props[P]' is not assignable to type 'Props[P]'.
          Type 'Props[P] | (({ query: string; } & { initializeFromUrl: any; })[P] extends Props[P] ? Props[P] : ({ query: string; } & { initializeFromUrl: any; })[P])' is not assignable to type 'Props[P]'.
            Type '({ query: string; } & { initializeFromUrl: any; })[P] extends Props[P] ? Props[P] : ({ query: string; } & { initializeFromUrl: any; })[P]' is not assignable to type 'Props[P]'.
              Type '({ query: string; } & { initializeFromUrl: any; })[P] | Props[P]' is not assignable to type 'Props[P]'.
                Type '({ query: string; } & { initializeFromUrl: any; })[P]' is not assignable to type 'Props[P]'.
                  Type '{ query: string; } & { initializeFromUrl: any; }' is not assignable to type 'Props'.
                    Type '{ query: string; } & { initializeFromUrl: any; }' is not assignable to type 'P'.
                      Type 'Props["onUpdateUrl"] | (string extends Props["query"] ? Props["query"] : string) | (any extends Props["initializeFromUrl"] ? Props["initializeFromUrl"] : any) | (keyof P extends "query" | "initializeFromUrl" ? ({ ...; } & { ...; })[("query" & keyof P) | ("initializeFromUrl" & keyof P)] extends Props[("query" & keyof...' is not assignable to type 'Props[P]'.
                        Type 'P["onUpdateUrl"] & ((queryString: string) => void)' is not assignable to type 'Props[P]'.ts(2345)
Peek Problem
No quick fixes available

```